src/Fcf.hs

```diff
@@ -302,6 +302,18 @@ data Not :: Bool -> Exp Bool
 type instance Eval (Not 'True)  = 'False
 type instance Eval (Not 'False) = 'True
 
+type Otherwise = 'True
+
+data Guarded :: [(Bool, Exp a)] -> Exp a
+type instance Eval (Guarded ('(p, x) ': xs)) =
+    If p (Eval x) (Eval (Guarded xs))
```

Comment:
Commuting `Eval` and `If` (so `Eval` is on the outside) would make this lazier. Currently, if `p` and `Eval x` get stuck, `Eval (Guarded xs)` will be evaluated.

Reply:
I had overlooked this, thanks!